src/katsdpmodels/primary_beam.py

```diff
@@ -234,7 +323,437 @@ def sample_grid(self, l: ArrayLike, m: ArrayLike, frequency: u.Quantity,   # noq
+        # region. If not (common case) we can avoid computing masks.
+        max_l = np.max(np.abs(l))
+        max_m = np.max(np.abs(m))
+        max_wavenumber = np.max(wavenumber)
+        limit_l = 0.5 / abs(self.x_step)
+        limit_m = 0.5 / abs(self.y_step)
+        if max_l * max_wavenumber > limit_l:
+            invalid = np.multiply.outer(wavenumber, np.abs(l)) > limit_l
+            # Insert axes for m and Jones terms
+            invalid = invalid[..., np.newaxis, :, np.newaxis, np.newaxis]
+            np.copyto(out, np.nan, where=invalid)
+        if max_m * max_wavenumber > limit_m:
+            invalid = np.multiply.outer(wavenumber, np.abs(m)) > limit_m
+            # Insert axes for l and Jones terms
+            invalid = invalid[..., np.newaxis, np.newaxis, np.newaxis]
+            np.copyto(out, np.nan, where=invalid)
+
+        return out
+
+    def sample_grid(self, l: ArrayLike, m: ArrayLike, frequency: u.Quantity,   # noqa: E741
+                    frame: Union[AltAzFrame, RADecFrame],
+                    output_type: OutputType, *,
+                    out: Optional[np.ndarray] = None) -> np.ndarray:
+        # This is a completely unoptimised placeholder implementation.
+        l_ = _asarray(l, np.float32)
+        m_ = _asarray(m, np.float32)
+        if l_.ndim != 1 or m_.ndim != 1:
+            raise ValueError('l and m must be 1D')
+        if (not isinstance(frame, AltAzFrame)
+                or output_type not in {OutputType.JONES_HV, OutputType.UNPOLARIZED_POWER}):
+            # Can't take advantage of separability when applying an arbitrary
+            # parallactic angle rotation. However, there may be some value in
+            # supporting a fast path when PA is a multiple of pi/2, for users
+            # that will interpolate from that grid to other parallactic
+            # angles but don't want to deal with the handedness change.
+            return self.sample(
+                l_[np.newaxis, :], m_[:, np.newaxis], frequency,
+                frame, output_type, out=out)
+        else:
+            _check_out(out, output_type)
+            if output_type == OutputType.JONES_HV:
+                return self._sample_grid_altaz_jones(l_, m_, frequency, out=out)
+            else:
+                mid = self._sample_grid_altaz_jones(l_, m_, frequency)
+                return self._finalize(mid, output_type, out=out)
+
+    @classmethod
+    def from_hdf5(cls: Type[_P], hdf5: h5py.File) -> _P:
+        samples = models.get_hdf5_dataset(hdf5, 'aperture_plane')
+        samples = models.require_columns('aperture_plane', samples, np.complex64, 5)
+        frequency = models.get_hdf5_dataset(hdf5, 'frequency')
+        frequency = models.require_columns('frequency', frequency, np.float32, 1)
+        # Quantity doesn't play nice with h5py numpy-like's, so force loading
+        frequency = frequency[:]
+        frequency <<= u.Hz
```

Comment:
self note: is bitwise left shift used instead of multiplication for performance reasons? Where else is and should it be used?

Reply:
The difference with multiplication is that it doesn't make a copy of the underlying data.

Reply:
https://docs.astropy.org/en/stable/units/#performance-tips